### ISSUE TITLE
Investigate goroutine leakage issue in Kubelet

### ISSUE THREAD

**Issue by dchen1107** [kubernetes/kubernetes#15805]
I profile kubelet without much workload (one or two pods), and found both goroutine and heap increased crazily. For example, on node:

/debug/pprof/

profiles:
0	block
1010	goroutine
1658	heap
12	threadcreate

And still keeping going. 

Among 1010 goroutines: there are >450 with same stack trace:

```
goroutine 100 [IO wait]:
net.(*pollDesc).Wait(0xc2081e6a00, 0x72, 0x0, 0x0)
	/usr/src/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc2081e6a00, 0x0, 0x0)
	/usr/src/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).Read(0xc2081e69a0, 0xc20887e000, 0x1000, 0x1000, 0x0, 0x7f78d4509048, 0xc2096c7ea8)
	/usr/src/go/src/net/fd_unix.go:242 +0x40f
net.(*conn).Read(0xc20803a788, 0xc20887e000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/src/go/src/net/net.go:121 +0xdc
crypto/tls.(*block).readFromUntil(0xc20829c1b0, 0x7f78d4516f50, 0xc20803a788, 0x5, 0x0, 0x0)
	/usr/src/go/src/crypto/tls/conn.go:454 +0xe6
crypto/tls.(*Conn).readRecord(0xc2080a5b80, 0x17, 0x0, 0x0)
	/usr/src/go/src/crypto/tls/conn.go:539 +0x2da
crypto/tls.(*Conn).Read(0xc2080a5b80, 0xc208260000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/src/go/src/crypto/tls/conn.go:904 +0x166
net/http.noteEOFReader.Read(0x7f78d451dcc8, 0xc2080a5b80, 0xc20839e5d8, 0xc208260000, 0x1000, 0x1000, 0x1540420, 0x0, 0x0)
	/usr/src/go/src/net/http/transport.go:1270 +0x6e
net/http.(*noteEOFReader).Read(0xc20836fa80, 0xc208260000, 0x1000, 0x1000, 0xc20813a900, 0x0, 0x0)
	<autogenerated>:125 +0xd4
bufio.(*Reader).fill(0xc20824be00)
	/usr/src/go/src/bufio/bufio.go:97 +0x1ce
bufio.(*Reader).Peek(0xc20824be00, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/src/go/src/bufio/bufio.go:132 +0xf0
net/http.(*persistConn).readLoop(0xc20839e580)
	/usr/src/go/src/net/http/transport.go:842 +0xa4
created by net/http.(*Transport).dialConn
	/usr/src/go/src/net/http/transport.go:660 +0xc9f
```
or

```
goroutine 36 [select, 76 minutes]:
net/http.(*persistConn).readLoop(0xc20839e000)
	/usr/src/go/src/net/http/transport.go:928 +0x9ce
created by net/http.(*Transport).dialConn
	/usr/src/go/src/net/http/transport.go:660 +0xc9f
```

Then another 477 with net/http.(*persistConn).writeLoop

**Comment by yujuhong** [comment]
Similar goroutine leakage from my cluster that had been running for 5 hours. My kubelet happily accumulate over 18000 goroutines.

```
8930 @ 0x417025 0x4138e0 0x412d78 0x56a587 0x56a5f3 0x56c2af 0x57f75c 0x5c6b0e 0x5d0a14 0x7a59fe 0x7a5c40 0x5c4304 0x440751
#	0x56a587	net.(*pollDesc).Wait+0x47		/usr/src/go/src/net/fd_poll_runtime.go:84
#	0x56a5f3	net.(*pollDesc).WaitRead+0x43		/usr/src/go/src/net/fd_poll_runtime.go:89
#	0x56c2af	net.(*netFD).Read+0x40f			/usr/src/go/src/net/fd_unix.go:242
#	0x57f75c	net.(*conn).Read+0xdc			/usr/src/go/src/net/net.go:121
#	0x5c6b0e	net/http.noteEOFReader.Read+0x6e	/usr/src/go/src/net/http/transport.go:1270
#	0x5d0a14	net/http.(*noteEOFReader).Read+0xd4	<autogenerated>:125
#	0x7a59fe	bufio.(*Reader).fill+0x1ce		/usr/src/go/src/bufio/bufio.go:97
#	0x7a5c40	bufio.(*Reader).Peek+0xf0		/usr/src/go/src/bufio/bufio.go:132
#	0x5c4304	net/http.(*persistConn).readLoop+0xa4	/usr/src/go/src/net/http/transport.go:842

8935 @ 0x417025 0x4192cc 0x418762 0x5c540d 0x440751
#	0x5c540d	net/http.(*persistConn).writeLoop+0x41d	/usr/src/go/src/net/http/transport.go:945
```

**Comment by dchen1107** [comment]
As an record: I measured this before 1.0 release, and found some leakages from cAdvsor. After that fix, we reasonably keep the goroutine down to some number (i.e. 100)

This should be a regression. This explains Kubelet CPU, memory usage increasing.

**Comment by yujuhong** [comment]
I found the culprit.

TL;DR: go-dockerclient now saves and reuses the connection. `cadvisor.GetVersionInfo` creates a new docker client each time, causing the connection leakage. This does not affect v1.1 because go-dockerclient was only updated recently (#15127).

Kubelet calls cadvisor to get version info [whenever updating the node status](https://github.com/kubernetes/kubernetes/blob/0338e0ef5383037dcba5807ac8047c6b306e14c3/pkg/kubelet/kubelet.go#L2375). In every `cadvisor.GetVersionInfo` function call, cadvisor [creates and using a new docker client each time] (https://github.com/google/cadvisor/blob/0.16.0.2/manager/machine.go#L146) to get the Docker version.

A recent go-dockerclient commit https://github.com/fsouza/go-dockerclient/commit/163268693e2cf8be2920158b59ef438fc77b85e2 changes to use http.Client for HTTP requests over UNIX domain socket. This changes the behavior of the client - it no longer closes the connection immediately after the request, instead, it [saves it for future use] (https://github.com/nitrous-io/go-dockerclient/commit/163268693e2cf8be2920158b59ef438fc77b85e2#diff-d0b7f86cea13f90d3cc0cc2338da8d5fR678).

Note that the cadvisor repository still uses the older go-dockerclient, hence could not have caught the leakage.

Removing this issue from v1.1milestone since it only affects HEAD (verified).

Simple code to reproduce this issue:
Using cadvisor (through k8s): https://gist.github.com/yujuhong/bed3f6757da772bef50b
Using go-dockerclient directly: https://gist.github.com/yujuhong/6068441fb45746dfcc77

**Comment by pmorie** [comment]
Classy digging :trophy:

**Comment by vishh** [comment]
Some PRs from cadvisor ([#909](https://github.com/google/cadvisor/pull/909), @jimmidyson's prometheus PRs) and a pending PR for issue [#920](https://github.com/google/cadvisor/issues/920), might have to be cherry-picked into v1.1. I think resolving this issue right away in cadvisor is necessary.

**Comment by yujuhong** [comment]
> Some PRs from cadvisor (#909, @jimmidyson's prometheus PRs) and a pending PR for issue #920, might have to be cherry-picked into v1.1. I think resolving this issue right away in cadvisor is necessary.

We should resolve this issue in cadvisor, but as long as we don't bump the go-dockerclient version in v1.1, v1.1 wouldn't be affected.

**Comment by yujuhong** [comment]
Update: cadvisor also creates a new docker client for each container. With the new go-dockerclient, it would use one connection per container, and never reclaims those connections after the containers die.

**Comment by dchen1107** [comment]
Here are the reason I moved this back to v1.1 milestone:

https://github.com/kubernetes/kubernetes/pull/15612#issuecomment-148404089

cc/ @pmorie Do we really need AdditionalGroups feature for 1.1 release? Otherwise, we don't need to bump go-dockerclient library to include  commit fsouza/go-dockerclient@1632686

**Comment by dchen1107** [comment]
All pr required from cAdvisor were merged, we just need to cut a new release: https://github.com/google/cadvisor/issues/935, and update our GoDep here.

**Comment by pmorie** [comment]
@dchen1107 I would like to have SupplementalGroups for use in 1.1, yeah.

**Comment by paralin** [comment]
Is this what causes my Kubelets to use a massive amount of memory?

**Comment by dchen1107** [comment]
@paralin If you are running kubelet from HEAD, you might hit by this issue caused by misuse of new go-dockerclient library by cAdvisor. But if you are running 1.0 release, that should be a different issue.

**Comment by paralin** [comment]
@dchen1107 It's on Head. Odd.

**Comment by yujuhong** [comment]
@paralin: The fix #16127 (cadvisor version bump) should be merged soon.

**Comment by bgrant0607** [comment]
Needs cherrypick, also
